### ISSUE TITLE
StorageSpec for prometheus's persistent volume is not picked up by prometheus-operator

### ISSUE THREAD

**Issue by MIllgner** [coreos/prometheus-operator#1121]
**What did you do?**
I tried to configure prometheus-operator's value file to get a PVC/PV instead of an emptyDir{}. I did it with following commands:

helm upgrade -i prometheus-operator coreos/prometheus-operator  --namespace monitoring
helm upgrade -i kube-prometheus coreos/kube-prometheus  --namespace monitoring -f values.yaml

**What did you expect to see?**

I expected to get a PVC and a PV and a kube-prometheus pod which is picking up this volume. The value file contains:
```
storageSpec:
    volumeClaimTemplate:
      spec:
        storageClassName: ibmc-file-gold
        accessModes: ["ReadWriteMany"]
        resources:
          requests:
            storage: 50Gi
      selector:
        matchLabels:
          app: prometheus
```

**What did you see instead? Under which circumstances?**

I got a failure during deployment:
```
Error: UPGRADE FAILED: Prometheus.monitoring.coreos.com "kube-prometheus" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"...
spec.storage.class in body is required
spec.storage.selector in body is required
spec.storage.resources in body is required
```
Then I added the requested values in the values.yaml file and deployed:
```
storageSpec:
    class: ""
    selector:
      matchLabels:
        app: prometheus
    resources: {}
    volumeClaimTemplate:
      spec:
        storageClassName: ibmc-file-gold
        accessModes: ["ReadWriteMany"]
        resources:
          requests:
            storage: 50Gi
      selector:
        matchLabels:
          app: prometheus
```
With this configuration the error message went away ... BUT

There is no change for kube-prometheus POD. It still has the emptyDir as data volume.
And in addition no PVC / PV is created.

**Environment**

* Kubernetes version information:

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.4", GitCommit:"bee2d1505c4fe820744d26d41ecd3fdd4a3d6546", GitTreeState:"clean", BuildDate:"2018-03-12T16:29:47Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.3-2+0b1ffe38401940", GitCommit:"0b1ffe38401940a979b81c581f6774888c1ead7f", GitTreeState:"clean", BuildDate:"2018-03-14T08:29:35Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by gianrubio** [comment]
It happened the same to me yesterday but I forgot to fill an issue, for some reason the upgrade in the `storageSpec` in the Prometheus `crd` are not triggering a change to the statefulSet of prometheus. 

@MIllgner could you provide logs from prometheus-operator when you added the storageSpec?

**Comment by MIllgner** [comment]
@gianrubio I see following error in the prometheus-operator log:

```
E0321 09:03:24.899446       1 operator.go:579] Sync "monitoring/kube-prometheus" failed: updating statefulset failed: StatefulSet.apps "prometheus-kube-prometheus" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden.
```

**Comment by brancz** [comment]
This is not a bug, it's just that storage cannot be changed after it was initially created. What we could probably do is that we recreate on any changes to an emptydir volume or a change from emptydir -> storage, but we can't modify any storage that is already in place.

**Comment by brancz** [comment]
For now you need to re-create the Prometheus object, we'll have to clarify why this happens in the StatefulSet controller, and if it makes any sense to override this behavior on our side, or if possible just fix things upstream.

**Comment by richerve** [comment]
Meanwhile you can do `kubectl delete sts prometheus-kube-prometheus` and in the next loop the new StatefulSet will be picked up and the volume configured. This will not delete the Pods but if you have only one replica the service will be down for a moment.

**Comment by mr-prud** [comment]
I think is a bug, because even in creation the prometheus-operator api invalid request when 
` spec.storage.class in body is required`
` spec.storage.selector in body is required`
` spec.storage.resources in body is required`
Actually we must set spec.storage.[class,selector,resources] with empty value to be "api compliant" but this value are not used.
the storageSpec.volumeClaimTemplate is sufficient to deploy a working statefulset

**Comment by brancz** [comment]
yeah that's a bug as described in #992 we're soon gonna be releasing a new version that will fix this

**Comment by MIllgner** [comment]
@brancz @richerve @mr-prud Thanks for the explanation. I deleted the statefulSet and now the storageSpec is picked up by prometheus-operator. When the forementioned bug is fixed, I will remove the depricated values too. 
By the way, I tried to follow prometheus' s storage users guide:
https://github.com/coreos/prometheus-operator/blob/master/Documentation/user-guides/storage.md
which was not really helpful in that specific situation.
